File: Example/FWFramework/Classes/Module/Test/Module/TestPopupController.m
//
//  TestPopupController.m
//  FWFramework_Example
//
//  Created by wuyong on 2022/8/25.
//  Copyright © 2022 CocoaPods. All rights reserved.
//

#import "TestPopupController.h"
#import "AppSwift.h"
@import FWFramework;

#define TITLES @[@"修改", @"删除", @"扫一扫", @"付款"]
#define ICON_PHONE FWIconImage(@"zmdi-var-smartphone-iphone", 24)
#define ICONS  @[FWIconImage(@"zmdi-var-edit", 24), FWIconImage(@"zmdi-var-delete", 24), ICON_PHONE, FWIconImage(@"zmdi-var-card", 24)]

@interface TestPopupController () <FWViewController, FWPopupMenuDelegate, UITextFieldDelegate>

@property (weak, nonatomic) UITextField *textField;
@property (weak, nonatomic) UILabel *customCellView;

@property (nonatomic, strong) FWPopupMenu *popupMenu;

@end

@implementation TestPopupController

- (void)setupSubviews
{
    UIButton *button = [UIButton fw_buttonWithImage:ICON_PHONE];
    [button fw_addTouchTarget:self action:@selector(onPopupClick:)];
    [self.view addSubview:button];
    button.fw_layoutChain.leftWithInset(10).topToSafeAreaWithInset(10).size(CGSizeMake(44, 44));
    
    button = [UIButton fw_buttonWithImage:ICON_PHONE];
    [button fw_addTouchTarget:self action:@selector(onPopupClick:)];
    [self.view addSubview:button];
    button.fw_layoutChain.rightWithInset(10).topToSafeAreaWithInset(10).size(CGSizeMake(44, 44));
    
    button = [UIButton fw_buttonWithImage:ICON_PHONE];
    [button fw_addTouchTarget:self action:@selector(onPopupClick:)];
    [self.view addSubview:button];
    button.fw_layoutChain.leftWithInset(10).bottomWithInset(10).size(CGSizeMake(44, 44));
    
    button = [UIButton fw_buttonWithImage:ICON_PHONE];
    [button fw_addTouchTarget:self action:@selector(onPopupClick:)];
    [self.view addSubview:button];
    button.fw_layoutChain.rightWithInset(10).bottomWithInset(10).size(CGSizeMake(44, 44));
    
    UITextField *textField = [UITextField new];
    textField.placeholder = @"我是输入框";
    textField.textColor = [AppTheme textColor];
    [textField fw_setBorderColor:[AppTheme borderColor] width:0.5 cornerRadius:5];
    _textField = textField;
    textField.delegate = self;
    [self.view addSubview:textField];
    textField.fw_layoutChain.leftWithInset(50).rightWithInset(50).topToSafeAreaWithInset(200).height(45);
    
    UILabel *customLabel = [UILabel fw_labelWithFont:[UIFont fw_fontOfSize:16] textColor:[AppTheme textColor] text:@"我是自定义标签"];
    _customCellView = customLabel;
    customLabel.backgroundColor = [AppTheme cellColor];
    [self.view addSubview:customLabel];
    customLabel.fw_layoutChain.centerX().topToViewBottomWithOffset(textField, 50).size(CGSizeMake(200, 50));
    
    [self.view setNeedsLayout];
    [self.view layoutIfNeeded];
}

- (void)onPopupClick:(UIButton *)sender {
    [FWPopupMenu showRelyOnView:sender titles:TITLES icons:ICONS menuWidth:120 otherSettings:^(FWPopupMenu *popupMenu) {
        popupMenu.delegate = self;
    }];
}

- (void)onTestClick:(UIButton *)sender {
    [FWPopupMenu showRelyOnView:sender titles:@[@"111",@"222",@"333",@"444",@"555",@"666",@"777",@"888"] icons:nil menuWidth:100 otherSettings:^(FWPopupMenu *popupMenu) {
        popupMenu.priorityDirection = FWPopupMenuPriorityDirectionLeft;
        popupMenu.borderWidth = 1;
        popupMenu.borderColor = [UIColor redColor];
        popupMenu.arrowPosition = 22;
    }];
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
{
    UITouch *t = touches.anyObject;
    CGPoint p = [t locationInView: self.view.window];
    
    CGRect customFrame = [self.customCellView.superview convertRect:self.customCellView.frame toView:self.view.window];
    if (CGRectContainsPoint(customFrame, p)) {
        [self showCustomPopupMenuWithPoint:p];
    }else {
        [self showDarkPopupMenuWithPoint:p];
    }
}

- (void)showDarkPopupMenuWithPoint:(CGPoint)point
{
    [FWPopupMenu showAtPoint:point titles:TITLES icons:nil menuWidth:110 otherSettings:^(FWPopupMenu *popupMenu) {
        popupMenu.dismissOnSelected = NO;
        popupMenu.isShowShadow = YES;
        popupMenu.delegate = self;
        popupMenu.offset = 10;
        popupMenu.type = FWPopupMenuTypeDark;
        popupMenu.rectCorner = UIRectCornerBottomLeft | UIRectCornerBottomRight;
    }];
}

- (void)showCustomPopupMenuWithPoint:(CGPoint)point
{
    [FWPopupMenu showAtPoint:point titles:TITLES icons:nil menuWidth:110 otherSettings:^(FWPopupMenu *popupMenu) {
        popupMenu.dismissOnSelected = YES;
        popupMenu.isShowShadow = YES;
        popupMenu.delegate = self;
        popupMenu.type = FWPopupMenuTypeDefault;
        popupMenu.cornerRadius = 8;
        popupMenu.rectCorner = UIRectCornerTopLeft| UIRectCornerTopRight;
        popupMenu.tag = 100;
        //如果不加这句默认是 UITableViewCellSeparatorStyleNone 的
        popupMenu.tableView.separatorStyle = UITableViewCellSeparatorStyleSingleLine;
    }];
}

#pragma mark - FWPopupMenuDelegate
- (void)popupMenu:(FWPopupMenu *)popupMenu didSelectedAtIndex:(NSInteger)index
{
    //推荐回调
    NSLog(@"点击了 %@ 选项",popupMenu.titles[index]);
}

- (void)popupMenuBeganDismiss:(FWPopupMenu *)popupMenu
{
    if (self.textField.isFirstResponder) {
        [self.textField resignFirstResponder];
    }
}

- (UITableViewCell *)popupMenu:(FWPopupMenu *)popupMenu cellForRowAtIndex:(NSInteger)index
{
    if (popupMenu.tag != 100) {
        return nil;
    }
    static NSString * identifier = @"customCell";
    UITableViewCell * cell = [popupMenu.tableView dequeueReusableCellWithIdentifier:identifier];
    if (!cell) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"customCell"];
    }
    
    cell.textLabel.text = TITLES[index];
    cell.imageView.image = ICONS[index];
    return cell;
}

#pragma mark - UITextFieldDelegate
- (BOOL)textFieldShouldBeginEditing:(UITextField *)textField
{
    _popupMenu = [FWPopupMenu showRelyOnView:textField titles:@[@"密码必须为数字、大写字母、小写字母和特殊字符中至少三种的组合，长度不少于8且不大于20"] icons:nil menuWidth:textField.bounds.size.width otherSettings:^(FWPopupMenu *popupMenu) {
        popupMenu.delegate = self;
        popupMenu.showMaskView = NO;
        popupMenu.priorityDirection = FWPopupMenuPriorityDirectionBottom;
        popupMenu.maxVisibleCount = 1;
        popupMenu.itemHeight = 60;
        popupMenu.borderWidth = 1;
        popupMenu.fontSize = 12;
        popupMenu.dismissOnTouchOutside = YES;
        popupMenu.dismissOnSelected = NO;
        popupMenu.borderColor = [UIColor brownColor];
        popupMenu.textColor = [UIColor brownColor];
        popupMenu.animationManager.style = FWPopupMenuAnimationStyleFade;
        popupMenu.animationManager.duration = 0.15;
    }];
    return YES;
}

- (BOOL)textFieldShouldReturn:(UITextField *)textField
{
    [_popupMenu dismiss];
    return YES;
}

@end

File: Example/FWFramework/Classes/Module/Test/Module/TestPopupController.swift
//
//  TestPopupController.swift
//  FWFramework_Example
//
//  Created by wuyong on 2022/9/29.
//  Copyright © 2022 CocoaPods. All rights reserved.
//

import FWFramework

class TestPopupController: UIViewController, ViewControllerProtocol, PopupMenuDelegate, UITextFieldDelegate {
    let titles: [String] = ["修改", "删除", "扫一扫", "付款"]
    let iconPhone: UIImage = APP.iconImage("zmdi-var-smartphone-iphone", 24)!
    let icons: [UIImage] = [APP.iconImage("zmdi-var-edit", 24)!, APP.iconImage("zmdi-var-delete", 24)!, APP.iconImage("zmdi-var-smartphone-iphone", 24)!, APP.iconImage("zmdi-var-card", 24)!]

    var popupMenu: PopupMenu?

    private lazy var textField: UITextField = {
        let result = UITextField()
        result.placeholder = "我是输入框"
        result.textColor = AppTheme.textColor
        result.app.setBorderColor(AppTheme.borderColor, width: 0.5, cornerRadius: 5)
        result.delegate = self
        return result
    }()

    private lazy var customLabel: UILabel = {
        let result = UILabel.app.label(font: UIFont.app.font(ofSize: 16), textColor: AppTheme.textColor, text: "我是自定义标签")
        result.backgroundColor = AppTheme.cellColor
        return result
    }()

    private lazy var titleView: ToolbarTitleView = {
        let result = ToolbarTitleView()
        result.alignmentLeft = true
        result.isExpandedSize = true
        return result
    }()

    func setupNavbar() {
        navigationItem.titleView = titleView
        app.addRightBarItem(APP.iconImage("zmdi-var-help-outline", 24)) { [weak self] _ in
            self?.showGuide()
        }
        app.addRightBarItem(UIBarButtonItem.SystemItem.action) { [weak self] _ in
            let customView = UIImageView()
            customView.image = UIImage.app.appIconImage()
            customView.frame = CGRect(x: 0, y: 0, width: 100, height: 100)
            customView.isUserInteractionEnabled = true

            let popupMenu = PopupMenu.show(at: self?.view.center ?? .zero, customView: customView, menuWidth: customView.frame.width) { popupMenu in
                popupMenu.dismissOnTouchOutside = false
                popupMenu.arrowHeight = 0
            }
            customView.app.addTapGesture { _ in
                popupMenu.dismiss()
            }
        }
    }

    func setupSubviews() {
        var button = UIButton.app.button(image: iconPhone)
        button.app.addTouch(target: self, action: #selector(onPopupClick(_:)))
        view.addSubview(button)
        button.app.layoutChain
            .left(10)
            .top(toSafeArea: 10)
            .size(CGSize(width: 44, height: 44))

        button = UIButton.app.button(image: iconPhone)
        button.app.addTouch(target: self, action: #selector(onPopupClick(_:)))
        view.addSubview(button)
        button.app.layoutChain
            .right(10)
            .top(toSafeArea: 10)
            .size(CGSize(width: 44, height: 44))

        button = UIButton.app.button(image: iconPhone)
        button.app.addTouch(target: self, action: #selector(onPopupClick(_:)))
        view.addSubview(button)
        button.app.layoutChain
            .left(10)
            .bottom(10)
            .size(CGSize(width: 44, height: 44))

        button = UIButton.app.button(image: iconPhone)
        button.app.addTouch(target: self, action: #selector(onPopupClick(_:)))
        view.addSubview(button)
        button.app.layoutChain
            .right(10)
            .bottom(10)
            .size(CGSize(width: 44, height: 44))

        view.addSubview(textField)
        textField.app.layoutChain
            .left(50)
            .right(50)
            .top(toSafeArea: 200)
            .height(45)

        view.addSubview(customLabel)
        customLabel.app.layoutChain
            .centerX()
            .top(toViewBottom: textField, offset: 50)
            .size(CGSize(width: 200, height: 50))

        view.setNeedsLayout()
        view.layoutIfNeeded()
    }

    func showGuide() {
        var items: [GuideViewItem] = []
        let textItem = GuideViewItem(sourceView: textField, text: "我是输入框的引导")
        items.append(textItem)
        let labelItem = GuideViewItem(sourceView: customLabel, image: UIImage.app.appIconImage())
        items.append(labelItem)
        let rectItem = GuideViewItem(rect: CGRect(x: 10, y: app.topBarHeight + 10, width: 44, height: 44), text: NSAttributedString(string: "我是最左侧的引导", attributes: [.font: UIFont.app.font(ofSize: 16), .foregroundColor: UIColor.yellow]))
        items.append(rectItem)

        let vc = GuideViewController(items: items)
        vc.arrowImage = UIImage(named: "guideArrow")
        vc.indexWillChangeBlock = { index, _ in
            print("showGuide indexWillChangeBlock: \(index)")
        }
        vc.indexDidChangeBlock = { index, _ in
            print("showGuide indexDidChangeBlock: \(index)")
        }
        vc.show(from: self) {
            print("showGuide completion")
        }
    }

    @objc func onPopupClick(_ sender: UIButton) {
        PopupMenu.show(relyOn: sender, titles: titles, icons: icons, menuWidth: 120) { [weak self] popupMenu in
            popupMenu.delegate = self
        }
    }

    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        guard let touch = touches.randomElement() else { return }
        let point = touch.location(in: view.window)
        let frame = customLabel.superview?.convert(customLabel.frame, to: view.window) ?? .zero
        if frame.contains(point) {
            showCustomPopupMenu(point)
        } else {
            showDarkPopupMenu(point)
        }
    }

    private func showCustomPopupMenu(_ point: CGPoint) {
        PopupMenu.show(at: point, titles: titles, icons: nil, menuWidth: 110) { popupMenu in
            popupMenu.dismissOnSelected = true
            popupMenu.showsShadow = true
            popupMenu.delegate = self
            popupMenu.arrowHeight = 0
            popupMenu.separatorColor = UIColor.red
            popupMenu.separatorInsets = UIEdgeInsets(top: 0, left: 8, bottom: 0, right: 8)
            popupMenu.imageEdgeInsets = UIEdgeInsets(top: 0, left: -8, bottom: 0, right: 0)
            popupMenu.titleEdgeInsets = UIEdgeInsets(top: 0, left: -8, bottom: 0, right: 0)
            // popupMenu.tag = 100
            popupMenu.rectCorner = [.topLeft, .topRight]
        }
    }

    private func showDarkPopupMenu(_ point: CGPoint) {
        PopupMenu.show(at: point, titles: titles, icons: nil, menuWidth: 110) { popupMenu in
            popupMenu.dismissOnSelected = false
            popupMenu.showsShadow = true
            popupMenu.delegate = self
            popupMenu.offset = 10
            popupMenu.textColor = UIColor.lightGray
            popupMenu.menuBackgroundColor = UIColor(red: 0.25, green: 0.27, blue: 0.29, alpha: 1)
            popupMenu.rectCorner = [.bottomLeft, .bottomRight]
        }
    }

    func popupMenu(_ popupMenu: PopupMenu, didSelectAt index: Int) {
        app.showMessage(text: "点击了 \(popupMenu.titles[index])")
    }

    func popupMenuBeganDismiss(_ popupMenu: PopupMenu) {
        if textField.isFirstResponder {
            textField.resignFirstResponder()
        }
    }

    func popupMenu(_ popupMenu: PopupMenu, cellForRowAt index: Int) -> UITableViewCell? {
        if popupMenu.tag != 100 { return nil }

        var cell = popupMenu.tableView.dequeueReusableCell(withIdentifier: "cell")
        if cell == nil {
            cell = UITableViewCell(style: .default, reuseIdentifier: "cell")
        }
        cell?.textLabel?.text = titles[index]
        cell?.imageView?.image = icons[index]
        return cell
    }

    func textFieldShouldBeginEditing(_ textField: UITextField) -> Bool {
        popupMenu = PopupMenu.show(relyOn: textField, titles: ["密码必须为数字、大写字母、小写字母和特殊字符中至少三种的组合，长度不少于8且不大于20"], icons: nil, menuWidth: textField.bounds.width, customize: { popupMenu in
            popupMenu.delegate = self
            popupMenu.maskViewColor = .clear
            popupMenu.priorityDirection = .bottom
            popupMenu.maxVisibleCount = 1
            popupMenu.itemHeight = 60
            popupMenu.borderWidth = 1
            popupMenu.font = UIFont.systemFont(ofSize: 12)
            popupMenu.dismissOnTouchOutside = true
            popupMenu.dismissOnSelected = false
            popupMenu.borderColor = .brown
            popupMenu.textColor = .brown
            popupMenu.animationManager.style = .fade
            popupMenu.animationManager.duration = 0.25
        })
        return true
    }

    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        popupMenu?.dismiss()
        return true
    }
}
